### ISSUE TITLE
FK constraint violation

### ISSUE THREAD

**Issue by GoogleCodeExporter** [sagarswathi/h2database#5]
```
What steps will reproduce the problem?
1. Create a new database and execute this simple script:
drop table b if exists;
drop table a if exists;

create table a (
id integer identity not null,
code varchar(10) not null,
primary key(id)
);

create table b (
name varchar(100) not null,
a integer,
primary key(name),
foreign key(a) references a(id)
);

2. Now launch the two classes attached (they're interactive).
Allow DBTestA to create a new record in table A. Jump into DBTestB's 
console and have it create a new B referencing A. Commit the insertion.
Next, return to DBTestA's console and rollback the creation of the A 
record.

3. You'll be most pleased to discover that in table A there are no 
records, while in table B there's a record pointing to a phantom A!

What's sad, very sad is that this FK violation occurs in an environment 
where transactions are supposed to be isolated at READ_COMMITED level.
Moreover, in DBTestB you get a timeout exception due to a lock on table A 
if you uncomment the commented lines that read from A. So the claims 
regarding MVCC do not verify either (the lock was supposed to be SHARED 
according to the docs)!

What is the expected output? What do you see instead?
I expect an exception to be thrown when the B record is created (since 
the isolation is at READ_COMMITTED level) or at least upon commit (at a 
lower level of isolation).

What version of the product are you using? On what operating system?
Version 1.0.61, Linux.
```

Original issue reported on code.google.com by `sulf...@gmail.com` on 20 Nov 2007 at 5:28

**Comment by GoogleCodeExporter** [comment]
```
Did you have a look at the TestMVCC class? This format would avoid 
attachements, and
it would simplify adding the test code, and therefore it would simplify solving 
the
problem.
```

Original comment by `thomas.t...@gmail.com` on 20 Nov 2007 at 7:59
- Added labels: **Type-Enhancement**
- Removed labels: **Type-Defect**

**Comment by GoogleCodeExporter** [comment]
```
[deleted comment]
```

**Comment by GoogleCodeExporter** [comment]
```
Ok, I'll use the pattern in TestMVCC.java next time.

But my point is not about MVCC and exclusive table-level locking: I 
demonstrated a 
huge bug of data integrity. When have FK constraints become 'enhancements'?
```

Original comment by `sulf...@gmail.com` on 20 Nov 2007 at 8:50

**Comment by GoogleCodeExporter** [comment]
```
> But my point is not about MVCC and exclusive table-level locking

Sorry I thought it is about MVCC: The test cases uses MVCC=TRUE. Is it possible 
to
get the problem without using MVCC? 
```

Original comment by `thomas.t...@gmail.com` on 21 Nov 2007 at 9:18
- Added labels: **Type-Defect**
- Removed labels: **Type-Enhancement**

**Comment by GoogleCodeExporter** [comment]
```
Thanks for reporting this problem! I agree it's a bug. I didn't find it first 
because
with MVCC it doesn't happen. The problem is that without MVCC, referential 
integrity
checks don't lock the referenced table. This will be fixed for the next 
release. Here
is a simpler test case:

    private void testReferential() throws Exception {
        deleteDb("transaction");        
        Connection c1 = getConnection("transaction");
        c1.setAutoCommit(false);
        Statement s1 = c1.createStatement();
        s1.execute("drop table if exists a");
        s1.execute("drop table if exists b");
        s1.execute("create table a (id integer identity not null, code varchar(10)
not null, primary key(id))");
        s1.execute("create table b (name varchar(100) not null, a integer, primary
key(name), foreign key(a) references a(id))");
        Connection c2 = getConnection("transaction");
        c2.setAutoCommit(false);
        s1.executeUpdate("insert into A(code) values('un cod')");
        Statement s2 = c2.createStatement();
        try {
            s2.executeUpdate("insert into B values('un B', 1)");
            error("Unexpected success");
        } catch (SQLException e) {
            checkNotGeneralException(e);
        }
        c2.commit();
        c1.rollback();      
        c1.close();
        c2.close();
    }

```

Original comment by `thomas.t...@gmail.com` on 21 Nov 2007 at 10:17
- Changed state: **Accepted**
- Added labels: **Priority-High**
- Removed labels: **Priority-Medium**

**Comment by GoogleCodeExporter** [comment]
```
Actually the FK constraint violation happens both with MVCC=TRUE and without it 
- 
I've just tested it again.
```

Original comment by `sulf...@gmail.com` on 22 Nov 2007 at 5:47

**Comment by GoogleCodeExporter** [comment]
```
This problem is fixed in the latest release (1.0.62), and test cases have been 
added.

> Actually the FK constraint violation happens both with MVCC=TRUE and without 
it

According to my test, it happened when the first connection (the one that 
opened the
database) didn't use MVCC=TRUE, and the second did. However this was a bug as 
well:
The second connection (or - any additional connection when the database is 
already
open) shouldn't be allowed to use MVCC=TRUE. Only the first connection can. In 
the
latest release (1.0.62) this will throw an exception.
```

Original comment by `thomas.t...@gmail.com` on 25 Nov 2007 at 6:23
- Changed state: **Fixed**
